### PR TITLE
add remote pinning services config

### PR DESCRIPTION
**Pull request by petar**
This is needed for https://github.com/ipfs/go-ipfs/pull/7661

### CODE REVIEW

**Review by lidel — CHANGES_REQUESTED** [review]
This PR  seems to duplicate already existing https://github.com/ipfs/go-ipfs-config/pull/104, so either we need to update it to follow this structure:

```js
{ 
  Pinning: {
    RemoteServices: {
      "name" {
        ApiEndpoint: "https://pins.example.com/",
        ApiKey: "foobar", 
        Policies: ["mfs-root"]
      } 
    } 
  }
}
```

or rebase the #104 and close this one.

**Comment by petar** [comment]
@lidel @aschmahmann  This PR is what https://github.com/ipfs/go-ipfs/pull/7661 is using. The main difference here is that pinning services is a map[string]RemotePinService, as opposed to []RemotePinService. Any reason to prefer a slice?

**Comment by aschmahmann** [comment]
I don't have a strong preference for slice vs map, I think what @lidel was asking for was to do `Pinning.RemoteServices` instead of just `RemotePinServices`

**Comment by lidel** [comment]
@petar I don't feel strongly either way, but we already use map[string] approach in [`Gateway.PublicGateways`](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#gatewaypublicgateways) so if there is no good reason to go other way, I'd go with the same here. No need to duplicate name, I believe its enough to have it once, in key. 

(updated my comment above to use it too)

**Comment by petar** [comment]
@lidel @aschmahmann What is the consensus design for the Policy field? map[string]string? I am looking into it as I'll need it for the MFS pinning feature.

**Comment by lidel** [comment]
Initially we would have either none or only "all-in-mfs", but in the long run we want to be able to have more than one policy enabled (eg. "all-in-mfs" and "all-local-pins"),

so either []string or map[string]bool(?) would work, as long we can read them back via `pin remote service ls --policy`

```json
Policy: ["all-in-mfs", "all-local-pins"]
```

or

```js
Policy: {
  mirrorMFS: true
  mirrorLocalPins: true
}
```

I don't know go(-ipfs) well enough to feel strongly either way – @aschmahmann thoughts?
Which will be easier to wire up and maintain?

**Comment by aschmahmann** [comment]
I don't have particularly strong feelings here, but probably maps/structs.

I tried to look at what our other config options do, such as how the gateways have

```
ipfs config --json Gateway.PublicGateways '{
    "ipfs.io": {
      "UseSubdomains": false,
      "Paths": ["/ipfs", "/ipns", "/api"]
    }
  }'
```

This of course is no help since we've used both options inside of a single command. Nonetheless, I think the "UseSubdomains" boolean is a little closer to what we're looking for here since structs like this are "typed" and so a little stricter and simpler to work with.

It could also make it easier to override defaults. Since maps ~ structs this means that if go-ipfs supports two policies "MirrorMFS" and "MirrorLocalPins" then we can say "by default local pins are duplicated remotely" (not necessarily a good idea, but for the sake of argument) and so MirrorLocalPins will default to true, while MirrorMFS defaults to false. If the user only sets MirrorMFS as true then MirrorLocalPins will remain the default instead of if it was an array where the user might not be sure if they were overriding MirrorLocalPins as false. Note I'm open to being told "having any remote pin default other than `none` is insane so just use an array to force users to override them all."

**Comment by lidel** [comment]
For what its worth, I can't come up with any policy that we would want to enable by default. IMO user agency takes a priority here. There should be no policies enabled: user needs to consciously opt-in.

**Comment by petar** [comment]
@lidel @aschmahmann I took a first stab at the MFS policy flag. See what you think.

File: remotepin.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+type Pinning struct {
+	RemoteServices map[string]RemotePinningService
+}
+
+type RemotePinningService struct {
+	ApiEndpoint string
+	ApiKey      string
+	// PinMFS enables watching for changes in MFS and re-pinning the MFS root cid whenever a change occurs.
+	PinMFS bool
```
Comment (aschmahmann) [inline remotepin.go:17]: I think this should be:

```
Policies: {
  PinMFS Flag
}
```

Where flag is the config type defined https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#flag and used in this package.

These two changes give us:
1) Cleaner look for multiple policies
2) Can easily tell if the user has set a value

File: remotepin.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	RemoteServicesSelector = PinningTag + "." + RemoteServicesTag
+)
+
+type Pinning struct {
+	RemoteServices map[string]RemotePinningService
+}
+
+type RemotePinningService struct {
+	ApiEndpoint string
+	ApiKey      string
```
Comment (aschmahmann) [inline remotepin.go:15]: WDYT about moving these into an Api sub struct, it'll allow us to mess around with the fields here later with less noise?

**Review by aschmahmann — CHANGES_REQUESTED** [review]

**Comment by aschmahmann** [comment]
All MFS work has been moved to #116

File: remotepin.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	RemoteServices   map[string]RemotePinningService
+}
+
+type RemotePinningService struct {
+	Api RemotePinningServiceApi
+}
+
+type RemotePinningServiceApi struct {
+	ApiEndpoint string
+	ApiKey      string
```
Comment (lidel) [inline remotepin.go:20]: nit: perhaps we can now avoid `Api.Api*`  and rename them to `Endpoint` and `Key`?

**Review by lidel — APPROVED** [review]
LGTM, but we may want to simplify names (either now or when MFS PR lands)

File: remotepin.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package config
+
+const (
+	PinningTag             = "Pinning"
+	RemoteServicesTag      = "RemoteServices"
+	RemoteServicesSelector = PinningTag + "." + RemoteServicesTag
+)
+
+type Pinning struct {
+	MFSRepinInterval string // in ns, us, ms, s, m, h
```
Comment (petar) [inline remotepin.go:10]: remove this one too?

**Review by petar — COMMENTED** [review]
LGTM

File: remotepin.go (hunk 3 above)
Comment (aschmahmann) [reply remotepin.go]: Done

File: remotepin.go (hunk 4 above)
Comment (aschmahmann) [reply remotepin.go]: Done

**Review by aschmahmann — APPROVED** [review]

**Pull request merged**